core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -544,6 +545,8 @@ final SearchContext createContext(ShardSearchRequest request, @Nullable Engine.S
                 indexService,
                 indexShard, scriptService, pageCacheRecycler, bigArrays, threadPool.estimatedTimeInMillisCounter(), parseFieldMatcher,
                 defaultSearchTimeout, fetchPhase);
+        context.getQueryShardContext().setFieldStatsProvider(new FieldStatsProvider(engineSearcher, indexService.mapperService()));
```

Comment:
I don't think we can check `IndicesService#canCache(ShardSearchRequest request, SearchContext context)` here for two reasons:

1. Because the rewrite is run before `#parseSource(DefaultSearchContext, SearchSourceBuilder)`, fields in the searchContext such as the `size` have not yet been set so the method returns false even when the request should be cached
2. Because the request has not been rewritten the method may return false because it has `now` in the query even though that might be rewritten to remove the `now` and should therefore be cached.

Maybe before we decide whether we should do this optimisation I should run some basic benchmarks to determine how much of an impact the rewrite has on queries that would not actually be cached anyway (e.g. because size is non-zero)?